[Ready] The Virtual Reality Hub

Whoneedspacee opened the pull request:
## About The Pull Request

VR sleepers now lead to the VR Hub, a place with a bunch of portals that lead you to other areas, or just has chill stuff like a bar. Includes portals to all the old vr maps, the megafauna training simulator map, a soon to be minecraft map, and my crappy game runner where you run away from falling tiles and fight to be the last person to survive. Merges murderdome and syndicate trainer into the vr hub map so they spawn every time with vr. Recall portals now can have multiple exits to support the old vr waypoint system from the hub. Also moves a lot of stuff around in the files to make it more organized. There's now also a ghost role that can access VR as well with infinite spawns.

VR NOW HAS A CONFIG OPTION SO ENABLE IT OR VR WILL NOT WORK.

## Why It's Good For The Game

Being a ghost is boring for most people, it's nice to have a place for people to hang out rather than just leave the server. This is also a nice place where people can put gamemodes they're testing or other fun little minigames that can't be main gamemodes. VR currently is not used by really anyone so I don't see anyone getting bothered by this PR as well.

[vr hub vfinal i dont even care anymore whatever (current version showcase)](https://www.youtube.com/watch?v=3E484ekd540&feature=youtu.be)

## Changelog
:cl: Joe Berry
add: The VR hub is now open, VR sleepers now lead to the same main area every time, where you can access all of the VR content from.
add: A VR connection from another station has now opened, ghosts can now access VR as a ghost spawn.
add: A new VR game, runner, run endlessly away from the dark tiles forming behind you and be the last one standing.
tweak: The vr megafauna arena portals now have multiple exits on every corner so you don't spawn on top of a megafauna and die immediately as often now.
config: Added a virtual reality config option to enable or disable the VR hub map.
fix: Wormholes can no longer teleport to every single portal in the game.
/:cl:

Whoneedspacee committed:
Adds a vr hub to connect to all vr content (random away missions, minigames on the hub map)

Adds equipment to be set when you move through recall portals (for mining stuff etc)

Moves recall portal id setting to the actual code instead of being handled in the map

Fixed a bug where the recall spell was added even if the teleport failed

Removes landmarks from most of the other vr maps in favor of the recall portal system so people can easily move back and forth from the hub, keeps them in case someone wants to make a different vr sleeper that connects to another place

Adds a ghost spawner to virtual reality (later they will be able to be pinged to become ghost spawns,  meant to be a place to wait for stuff like that and pai stuff)

Whoneedspacee committed:
One way portals can now have multiple exits

Adds multiple exits to each of the virtual reality away mission maps

Adds multiple exits to each of the megafauna arenas so you're less likely to spawn on top of a megafauna

Adds a config option to enable the virtual reality hub map

Portals with equipment to add now delete the equipment you already have on you

Whoneedspacee committed:
adds recall equipment (equipment that is set when you recall to that portal)

adds better lighting to the main vr hub area

fixes up the vr hub

Whoneedspacee committed:
wat

Whoneedspacee committed:
more lighting changes for the main hub area

moves portal stuff to their own respective game files unless in the case of the hub which handles the old vr stuff

changes the mining area to a general area for vr game stuff

adds framework for a new vr game called runner, run endlessly as tiles fall into chasms behind you, you have jumpboots to jump over chasms occasionally and you want to be the last person alive, or if there's only one person just until you fall

Whoneedspacee committed:
adds different areas for each vr location in the hub

runner is now an actual working game, there are win conditions, people cannot enter after it has started, and you can play endless mode if you're alone, and it resets after the game ends

runner no longer kills you when you lose and instead just recalls you automatically back to the entry portal

minor map changes to the hub

Whoneedspacee committed:
merges murderdome and syndicate trainer into the vr hub map so they're always enabled with vr, snowdin is stuck as a standalone map due to its size

adds a new icon state to the murderdome area to make it stand out on the map editor

adds a telecomms system to the vr hub map

Tlaltecuhtli commented:
finally, are there any problems performance wise?

Mickyan commented:
why is the main hub so dark?

zxaber commented:
Because it's on the dark-net. _Duh_

Whoneedspacee commented:
> why is the main hub so dark?

cause i havent finished putting everything in

> finally, are there any problems performance wise?

not seemingly, atmos is really the only performance problem in ss13 in general. i'm not sure if there's any way to disable atmos movement in areas but i will if that's a thing. if it isn't i'll just try to avoid atmos everywhere i can. it seems loading multiple away mission maps is laggier than loading just one big map though so thats why i decided to go with this, if thats what you were wondering about.

Whoneedspacee commented:
also if you guys have any reccomendations for minigames or the main vr hub area or just anything that the portals could go to, let me know.

zxaber commented:
Long as you're allowing ghosts to respawn, you could move the CTF arena in.

MrDoomBringer commented:
![image](https://user-images.githubusercontent.com/29008542/53291815-5a809680-3787-11e9-929a-09e9332d502e.png)
solarium real real ending confirmed/?????

MrDoomBringer commented:
@Whoneedspacee I know centcom doesn't have any atmos calculations run on it, so maybe look at the code there?

Also - could we add some kind of indicator to show the number of active minds in VR? Maybe a status display that shows how many gamers are in the VR rooms

Mickyan commented:
>Also - could we add some kind of indicator to show the number of active minds in VR? Maybe a status display that shows how many gamers are in the VR rooms
This could be nice for giving entertainment screens a reason to exist, let people directly spectate other players in VR

ATH1909 commented:
Idea for a traitor item: Something that lets the VR ghost role dudes steal the real world bodies of people who are in VR. Maybe the mind transfer spell (have the traitor item add a book of mind transfer to the list of megafauna loot for the virtual megafauna?) would work (swap bodies with someone's avatar, then hit the eject from VR button?)? I wouldn't want it to stick around outside of VR, though.

ModDownloading commented:
Would it be possible to have a magic mirror in the hub? This would let people change their VR look and name and the mirror already has settings in place to prevent wizards from abusing the overpowered species so I doubt that the VR players would be able to. As long as the mirror doesn't have some kind of impact on the VR player's actual body on the station I don't see too much of a problem in adding one.

Whoneedspacee commented:
As fun as VR body swapping sounds, I probably won’t be doing it since I try to keep controversial things out of my PR’s. And I think some people would be mad if I added something like that, just because they specifically know how I am. It’s probably better than VR emagging just killing people though. If someone else wanted to do that, that’d be great. I will say that mindswapping bodies in VR does actually switch your real bodies already so it wouldn’t be too difficult to implement.

Appearance changes in VR is probably a good idea. I’ll add an area where you can customize your looks and name I guess.

CTF is a good idea, I might add my own version with some cool stuff like classes and such.

Spectating people in VR with monitors sounds like a good ide. Haven’t really done anything with that though so no promises. This is mostly meant to be something other people add to as all after it’s PRed though. It is by no means complete.

Kmc2000 commented:
Did the minecraft map make it in?

Whoneedspacee commented:
It will when I get around to it, I definitely haven’t forgotten about it. As I said this PR is a ways away from completion and I’m just looking for ideas which you guys seem to love doing.

ATH1909 commented:
Idea here: a VR terminals/consoles/portals within the VR hub that allow VR avatars to control holodeck mobs (which will still disappear as normal if they leave the holodeck (while its safeties are on) or if the holodeck simulation changes/turns off, as per usual).

We might want to add more holodeck mobs to the holodeck simulations if that idea becomes a thing. A medibot in the emergency medical simulation wouldn't be out of place. Maybe add some more simulations specifically designed to cater to this feature, such as:
A simulation with one of every type of bot in it
A simulation with some syndicate simplemobs in it (I wonder if a player-controlled syndicate mech pilot can still hijack mechs...)
etc.

Maybe add a pAI program that'd let them enter VR from anywhere?

What if VR ghosts got the reward of becoming a Real Boy (TM) if they successfully kill every type of VR megafauna at least once without dying? They'd use a special single-use item (that could be stolen from them by another VR ghost) and materialize naked in the holodeck or something.

OverlyGeneric commented:
> What if VR ghosts got the reward of becoming a Real Boy (TM) if they successfully kill every type of VR megafauna at least once without dying? They'd use a special single-use item (that could be stolen from them by another VR ghost) and materialize naked in the holodeck or something.

I don't think it would that would be a good idea but cosmetics would be nice for VR ghosts but I've already stated it before about what expert megafaunas should give in #42353.

If anything you could have a reward for ghosts for doing so with a very limited holoform, similar to Goonstation's ghostcritters, made by a VR headset and interactable with controllers. 

The only way to make it balanced without worrying about policy changes would be allowing it to only pick up tiny objects and other whitelisted items deemed harmless, able to be dissipated easily, and only able to speak in VR-speak.

But... this should be made in another PR.

Whoneedspacee commented:
No. Unless a traitor stupidly body swaps another ghost or something like that, there will be no resurrections for dead people. That's just stupid.

It doesn't really make sense in lore anyways for them to resurrect, they're not actually ghosts they're random people from off-station.

I don't doubt that someone, maybe quite like me will just run the megafauna over and over respawning over and over. It's the same problem with bubblegums items.

OverlyGeneric commented:
> No. Unless a traitor stupidly body swaps another ghost or something like that, there will be no resurrections for dead people. That's just stupid.

I never said it was a resurrection or a fun way to play.

 I just said it would be like the ghost critters back on Goonstation, maybe even limiting it to be only working on holopads that are underused. 

It would be incredibly tedious to do anything as a hologhost at that point it wouldn't even be worth bothering the living.

Shadowflame909 commented:
@Whoneedspacee Martial arts and magic has already been made VR only, if obtained VR.

It's like a new brain entirely, you cant take it out with you. If you learn it in VR.


So all one would really need to do, is have emagging the machine spawn it.

Would be pretty amazing

datorangebottle commented:
Dude I fucking love this PR. A+.

pizzatiger commented:
I still think the other language isnt needed. We already have other ghost roles that can speak english and talk to the crew. This VR ghost role will be no different. You could even allow people to keep their original name and have the lore be that it has a personality database of all crewmembers and that they don't remember anything stationside, just give them a slightly funnier text to make them different from non-ghost VR people

zxaber commented:
Maybe, but the idea of space-Koreans is funny.

MrDoomBringer commented:
@Whoneedspacee You got a discord, gamer? I feel like it'd be fun to work on the map for this

Whoneedspacee commented:
> @Whoneedspacee You got a discord, gamer? I feel like it'd be fun to work on the map for this

If you want to help me with the map I am 100% for it. Banned from the /tg/ discord but my discord is Space#0789 anyone is free to add me if they want to help or recommend something for this PR.

Jordie0608 commented:
This is testmerged on bagil.

Whoneedspacee committed:
Adds recall effects, things that recall portals can do to the recalling mob or in general when something recalls to them

Fixes up runner code

Every player in runner now has the pacifist trait (no throwing mobs, no attacking other mobs, pushing, etc)

Tiles now randomly fall in runner to speed up the game and make it more interesting for a singular player

You are now given your surviving time, and the remaining players when you lose in runner, as well as being told you win when you do

Recall effects now handle the players losing in runner

Whoneedspacee commented:
testmerged wut why

new version https://www.youtube.com/watch?v=kSY2WQxfRAc&feature=youtu.be

Jordie0608 commented:
I wanted to see if we'd get some gameplay feedback but all I got was a bug report that away mission loading broke. Maybe I was too prompt here, is this meant to have a WIP tag on it?

Whoneedspacee commented:
> I wanted to see if we'd get some gameplay feedback but all I got was a bug report that away mission loading broke. Maybe I was too prompt here, is this meant to have a WIP tag on it?

Virtual Reality needs a config option enabled. Should I put that somewhere in the PR I dunno.

Basically this PR moves most of the virtual reality away missions into this map, and is leaving away missions for snowdin and the gateway stuff. If you don't have virtual reality enabled, then the hub won't load and virtual reality will not work.

Whoneedspacee commented:
@Jordie0608 If you want I can have an actually testmergeable version done by tomorrow since I'm at lecture, I'd definitely be interested in seeing what people think since this is a pretty major change for ghosts.

FloranOtten commented:
Calling it here; There will be a PR nerfing megafauna in the near future, due to too many people being competent at killing megafauna.

Whoneedspacee commented:
Then I’ll just make them harder.

nervere commented:
I'm removing the testmerge for this because it breaks loading away missions for VR.
It says the world is "misconfigured" and I'm unable to manually load in any away missions.

Whoneedspacee committed:
fixes runner bug where jumping on a black tile directly wouldn't kill you so you could just stand there infinitely never dying.

adds some more hub things like cameras for the ai, cosmetics, etc.

vr people now spawn as pacifists and the status is removed depending on which game they enter and come back from.

Whoneedspacee commented:
needs config enabled thats the error, yes i know we talked about this i just want it to be out there

https://www.youtube.com/watch?v=3E484ekd540&feature=youtu.be

i dont even care whatever

Whoneedspacee commented:
i'm burnt out tbh, it works fine though.

ShizCalev commented:
Conflict.

KetasDotepasto commented:
@Whoneedspacee You still working on this?

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into vrbegood

Whoneedspacee committed:
why you no mapmerge last commit

Whoneedspacee commented:
no, does anything need to be done?

KetasDotepasto commented:
Well, there was this merge conflict you just fixed.

I've checked the VR-hub out on my private server. It seems that the holopad in the bar can call other holopads on the station. Is that intentional? Definitively seems like a nice gimmick.

Whoneedspacee commented:
not intentional on my part but i did like the idea of some station connectivity so this was my solution instead of doing something like the spectator mode someone wanted.

KetasDotepasto commented:
I like this PR and hope it gets merged soon.

Best of luck to you.

ModDownloading commented:
Is there actually somewhere to change your VR character's name/appearance yet? I know I have suggested it already but I didn't see one in the video and since we have off-station people having access to VR I figure it might be nice to have one. Clothing change room looks great though.

Whoneedspacee commented:
No. I thought about doing magic mirror or something like it but I didn’t want to put in the work.

Magic mirror itself has races that can break VR by phasing through walls etc. If someone wants to add something like this after it gets merged that’s perfectly fine though.

ModDownloading commented:
Maybe just an edited magic mirror that has the VR-incompatible species removed from it if it wouldn't be too difficult? I don't think we need to have any non-station species in VR, but the reason for my suggestion was mainly so if people are joining from off-station they don't have to play as some random-generated character (unless they join in as their normal character instead of a random one, but that might cause difficulties).

Whoneedspacee commented:
aghhhhh fine ill do it

ModDownloading commented:
Sorry if I was being a bother about it, but thank you.

Whoneedspacee committed:
adds lesser magic mirror

Whoneedspacee commented:
nah its fine if it was that easy then it really should've been added im just trying to work on my gamemode rn and its stressful

ExcessiveUseOfCobblestone commented on code/modules/awaymissions/mission_code/runner.dm:
```diff
@@ -0,0 +1,134 @@
+		var/mob/living/carbon/human/H = user
+		GLOB.vr_runner_players -= H
+		if(GLOB.vr_runner_active)
+			if(GLOB.vr_runner_players.len == 0)
```
```suggestion
			if(!GLOB.vr_runner_players.len)
```

ExcessiveUseOfCobblestone commented on code/modules/awaymissions/mission_code/runner.dm:
```diff
@@ -0,0 +1,134 @@
+		GLOB.vr_runner_players -= H
+		if(GLOB.vr_runner_active)
+			if(GLOB.vr_runner_players.len == 0)
+				to_chat(H, "<span class='notice'>You win! You survived for [(world.time - game_start_time) / 10] seconds.</span>")
```
```suggestion
				to_chat(H, "<span class='notice'>You win! You survived for [(world.time - game_start_time) / 10] second\s.</span>")
```

ExcessiveUseOfCobblestone commented on code/modules/awaymissions/mission_code/runner.dm:
```diff
@@ -0,0 +1,134 @@
+				to_chat(H, "<span class='notice'>You win! You survived for [(world.time - game_start_time) / 10] seconds.</span>")
+				end_game()
+			else
+				to_chat(H, "<span class='notice'>You survived for [(world.time - game_start_time) / 10] seconds. [GLOB.vr_runner_players.len] other player(s) remained.</span>")
```
```suggestion
				to_chat(H, "<span class='notice'>You survived for [(world.time - game_start_time) / 10] second\s. [GLOB.vr_runner_players.len] other player(s) remained.</span>")
```

ExcessiveUseOfCobblestone commented on code/modules/awaymissions/mission_code/runner.dm:
```diff
@@ -0,0 +1,134 @@
+				to_chat(H, "<span class='notice'>You survived for [(world.time - game_start_time) / 10] seconds. [GLOB.vr_runner_players.len] other player(s) remained.</span>")
+
+/obj/effect/portal/permanent/one_way/recall/pit_faller/proc/game_start_countdown(wait_seconds = 10)
+	game_starting = 1
```
is that a boo madame?!!!

Whoneedspacee replied:
thanks for not misgendering me

ExcessiveUseOfCobblestone commented on code/modules/awaymissions/mission_code/runner.dm:
```diff
@@ -0,0 +1,134 @@
+/obj/effect/portal/permanent/one_way/recall/pit_faller/proc/game_start_countdown(wait_seconds = 10)
+	game_starting = 1
+	for(var/seconds_remaining = wait_seconds to 1 step -1)
+		if(GLOB.vr_runner_players.len == 0)
```
```suggestion
		if(!GLOB.vr_runner_players.len)
```

ExcessiveUseOfCobblestone commented on code/modules/awaymissions/mission_code/runner.dm:
```diff
@@ -0,0 +1,134 @@
+		for(var/mob/living/carbon/human/H in GLOB.vr_runner_players)
+			to_chat(H, "<span class='notice'>Game starting in [seconds_remaining].</span>")
+		sleep(10)
+	if(GLOB.vr_runner_players.len > 0)
```
```suggestion
	if(GLOB.vr_runner_players.len)
```

ExcessiveUseOfCobblestone commented on code/modules/awaymissions/mission_code/runner.dm:
```diff
@@ -0,0 +1,134 @@
+/turf/open/indestructible/runner/Entered(atom/movable/A)
+	. = ..()
+	if(isliving(A) && GLOB.vr_runner_active)
+		if(color == COLOR_ALMOST_BLACK)
```
> It's a color check

Whoneedspacee replied:
doesn't matter since there's no way to get any color changing stuff in there because it equips and gets rid of your equipment suck my ass

ExcessiveUseOfCobblestone commented on code/modules/awaymissions/mission_code/runner.dm:
```diff
@@ -0,0 +1,134 @@
+		if(color == COLOR_ALMOST_BLACK)
+			if(locate(A) in GLOB.vr_runner_players)
+				var/datum/thrownthing/TT = A.throwing
+				if(!TT || TT && TT.dist_travelled + 1 >= TT.maxrange)
```
```suggestion
				if(!TT || TT?.dist_travelled + 1 >= TT.maxrange)
```

Whoneedspacee replied:
i don't even know what you're recommending here

Jordie0608 replied:
`TT?.dist_travelled` will only check `dist_travelled` if `TT` is not null.

Whoneedspacee replied:
neato

ExcessiveUseOfCobblestone reviewed (commented):
fix ur boos or ima come to ur house

ExcessiveUseOfCobblestone committed:
Update code/modules/awaymissions/mission_code/runner.dm

Co-Authored-By: Whoneedspacee <yougotreallyowned@gmail.com>

ExcessiveUseOfCobblestone committed:
Update code/modules/awaymissions/mission_code/runner.dm

Co-Authored-By: Whoneedspacee <yougotreallyowned@gmail.com>

ExcessiveUseOfCobblestone committed:
Update code/modules/awaymissions/mission_code/runner.dm

Co-Authored-By: Whoneedspacee <yougotreallyowned@gmail.com>

Whoneedspacee committed:
Update runner.dm

fixing those dank boolean values

Whoneedspacee committed:
Update runner.dm

ExcessiveUseOfCobblestone commented:
ok this looks ok

ExcessiveUseOfCobblestone committed:
Update code/modules/awaymissions/mission_code/runner.dm

Co-Authored-By: Whoneedspacee <yougotreallyowned@gmail.com>

ExcessiveUseOfCobblestone commented:
did that break it lol

Whoneedspacee commented:
dont forget to friggan enable the config

Fikou commented:
why snowdin portal borked

Whoneedspacee commented:
i can only assume other away missions are loading, snowdin is still its own away mission so i can’t do anything about that.

adamkad1 commented:
Plasmamemes enter VR without their suits resulting in a quick deth, maybe make them spawn with suits, or everyone spawn as human

actioninja commented:
Just saw an insane bug on bagil. The wormholes from the wormhole even were teleporting people to the VR hub.

Whoneedspacee commented:
https://github.com/tgstation/tgstation/blob/58dc6df4cd08443a3a6ee7f1fa85e15426af3c9f/code/modules/events/wormholes.dm#L52, how old is this code, it can literally teleport to any portal in the entire game, that's why this bug is happening. It should probably only teleport between other wormhole portals.

Whoneedspacee commented:
Also for the plasmemes bug, there's already a lesser mirror for race changing in VR, and it only appears to kill you once you enter another portal which gets rid of your suit. The games aren't built for plasmemes anyways, and if you spawn in as a ghost it doesn't spawn you in as a plasmeme. So it's only a problem where a living plasma man spawns in and doesn't change his race. You can quickly recall and get your suit back anyways if you accidentally enter. Overall I don't see it as a problem to be fixed, and I'd prefer if people could be plasmamen in the main area at least.

Whoneedspacee committed:
adds fire extinguishers to the misc item chest

fixes absolutely stupid wormhole portal bug

Whoneedspacee committed:
Merge branch 'vrbegood' of https://github.com/Whoneedspacee/tgstation into vrbegood

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into vrbegood

office chair merge conflicts are stupid

Whoneedspacee commented:
Should be good to merge. Testmerges went fine people seemed to enjoy it, and the only seemingly major bug was the wormholes entering vr which is now fixed. There are technically ways to get stuff in other areas and blow shit up but honestly that's just normal ss13 stuff and I think leaving it in makes it more fun.

optimumtact commented:
it doesn't compile due to path changes

optimumtact commented on code/modules/awaymissions/mission_code/runner.dm:
```diff
@@ -0,0 +1,134 @@
+		return
+	not_reset = FALSE
+	color = COLOR_ALMOST_BLACK
+	for(var/mob/living/carbon/human/H in GLOB.vr_runner_players)
```
are you for real?

just search the contents of the turf

optimumtact commented on code/modules/awaymissions/mission_code/vrhub.dm:
```diff
@@ -0,0 +1,169 @@
+/obj/effect/mob_spawn/human/virtual_reality/equip(mob/living/carbon/human/H)
+	. = ..()
+	if(pacifist)
+		H.add_trait(TRAIT_PACIFISM, TRAUMA_TRAIT)
```
why is this not just behaviour on the mob_spawn/human path

optimumtact commented on code/modules/awaymissions/mission_code/vrhub.dm:
```diff
@@ -0,0 +1,169 @@
+	. = ..()
+	if(. && ishuman(M))
+		var/mob/living/carbon/human/H = M
+		H.remove_trait(TRAIT_PACIFISM, TRAUMA_TRAIT)
```
the second argument to the remove_trait proc is a list of sources, how is this valid code

More importantly, you should be adding a new source for this trait, it hasn't come from TRAUMA, it's coming from portals and should have a new define source added.

Whoneedspacee replied:
literally every other instance of remove_trait i've seen in the code doesn't use it as a list

optimumtact replied:
the proc definition definitely takes a list, i'll have to look at the logic and see what it does

optimumtact replied:
apparently it has a check in there
```
if(!islist(sources))
        sources = list(sources)
```

Whoneedspacee committed:
adds vr zone trait for special area effects like pacifism

vr areas can now instantly kill people, or toggle their pacifism instead of being handled through portals. allows for greater flexibility in the same areas and makes it easier to know what is supposed to be happening code wise.

the walls of the vr hub are now actually all indestructible making it all less likely to be destroyed to shit.

adds vr hub light emitters so all the lights cant be destroyed by annoying pranksters

adds a boxing ring type area to the vr hub where people can fight

optimumtact commented:
![joe fucking berry](https://file.house/iNqW.png)

Whoneedspacee committed:
im retarded

also fixes a bug with runner i guess

Whoneedspacee commented:
weird ghost bug fixed btw if you missed the memo

Whoneedspacee commented:
damnit

will fix when i get home as well as finish up by making the syndicate portal a portal that conpletely respawns you upon recalling

this is your fault you damn vr powergamers you ruined the fun for everyone else by going on murder sprees

on another note might add a jokey solarium puzzle now that the syndicate stuff cant break out

if you have any last requests for the main hub area for game stuff or quality of life you have about 9 hours to do so and then im not touching this ever again

Whoneedspacee committed:
ghosts that spawn in as vr people now ghost when they use the leave vr button

the syndicate portal is now one-way

Whoneedspacee committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into vrbegood

Whoneedspacee committed:
the syndicate portal now dusts you on an attempted recall because of stupid smuggling items

Whoneedspacee commented:
done 100% done to the done done done never touching again done

JJRcop commented on code/modules/VR/vr_human.dm:
```diff
@@ -33,6 +40,23 @@
+	if(A.death)
+		to_chat(src, "<span class='userdanger'>It is unwise to attempt to break Virtual Reality.</span>")
+		playsound(src, 'sound/effects/supermatter.ogg', 50, 1)
+		src.dust()
```
Just `dust()`

JJRcop commented on code/modules/VR/vr_human.dm:
```diff
@@ -33,6 +40,23 @@
+		src.add_trait(TRAIT_PACIFISM, VR_ZONE_TRAIT)
+		to_chat(src, "<span class='notice'>You feel like your ability to fight other living beings is being suppressed.</span>")
+	else if(!A.pacifist && src.has_trait(TRAIT_PACIFISM, VR_ZONE_TRAIT))
+		src.remove_trait(TRAIT_PACIFISM, VR_ZONE_TRAIT)
```
All of these should be `has_trait` or `remove_trait`, no `src.`

JJRcop commented on code/modules/VR/vr_human.dm:
```diff
@@ -44,6 +68,8 @@
 		vr_sleeper.vr_human = null
 		vr_sleeper = null
+	if(!real_mind && !vr_sleeper)
+		src.ghostize()
```
Just `ghostize()`

JJRcop commented on code/modules/awaymissions/mission_code/runner.dm:
```diff
@@ -0,0 +1,129 @@
+		return
+	not_reset = FALSE
+	color = COLOR_ALMOST_BLACK
+	for(var/mob/living/carbon/human/H in src.contents)
```
Just `contents`

JJRcop reviewed (commented):
Unnecessary `src.` usage

Whoneedspacee committed:
Update vr_human.dm

Whoneedspacee committed:
Update runner.dm

optimumtact commented on code/modules/VR/vr_human.dm:
```diff
@@ -33,6 +40,23 @@
+	if(A.pacifist && !has_trait(TRAIT_PACIFISM, VR_ZONE_TRAIT))
+		add_trait(TRAIT_PACIFISM, VR_ZONE_TRAIT)
+		to_chat(src, "<span class='notice'>You feel like your ability to fight other living beings is being suppressed.</span>")
+	else if(!A.pacifist && has_trait(TRAIT_PACIFISM, VR_ZONE_TRAIT))
```
why is there both a var and a trait?

optimumtact replied:
oh I see, this is on the area?
